userland/libutest/utest.c

```diff
@@ -0,0 +1,393 @@
+}
+
+/** \brief Notify the test supervisor, and await approval to continue the test.
+ */
+static void sync_with_supervisor(int svc) {
+    done = false;
+    ipc_notify_svc(svc);
+    yield_for(&done);
+}
+
+/** \brief Run a sequence of unit tests and report the results.
+ *
+ * This function is called by the IPC clients, i.e. the 'test runners'. This
+ * function coordinates with the test supervisor's IPC service to run each test
+ * in sequence and report the status of each one.
+ *
+ * \param tests An array of boolean functions which return true for PASS and
+ *              false for FAIL.
+ * \param test_count The total number of tests in the tests array.
+ * \param timeout_ms The maximum amount of time each test is allowed to run
+ *                   before being timed out.
+ * \param svc_name The IPC service name of the test supervisor (e.g.
+ *                 "org.tockos.utest")
+ */
+void utest_runner(utest_fun *tests, uint32_t test_count,
+                  uint32_t timeout_ms, const char *svc_name) {
+
+    // Initialize the test state.
+    memset(&test_buf[0], 0, TEST_BUF_SZ);
+    utest_t *test = (utest_t *)(&test_buf[0]);
+    test->count = test_count;
+    test->timeout_ms = timeout_ms;
+
+    // Establish communication with the test supervisor service.
+    int test_svc = ipc_discover(svc_name);
+    if (test_svc < 0) return;
+
+    // Register the callback for cooperative scheduling.
+    ipc_register_client_cb(test_svc, continue_cb, NULL);
+
+    // Share the test state with the supervisor.
+    ipc_share(test_svc, &test_buf[0], TEST_BUF_SZ);
+
+    // Wait for the supervisor's approval to start running tests.
+    test->cmd = TestInit;
+    sync_with_supervisor(test_svc);
+
+    uint32_t i = 0;
+    for (i = 0; i < test_count; i++) {
+        // Await approval to start the current test.
+        test->cmd = TestStart;
+        sync_with_supervisor(test_svc);
+
+        // Run the test.
+        bool passed = tests[i]();
+
+        // Record the result. If the test timed out, the supervisor will have
+        // marked the result already.
+        if (test->result != Timeout) {
+            test->result = passed ? Passed : Failed;
+        }
+
+        // Indicate test completion.
+        test->cmd = TestEnd;
+        sync_with_supervisor(test_svc);
+
+        // Move to the next test in the series.
+        test->current++;
+    }
+
+    // Indicate that the tests are all complete, and the app is cleaning up.
+    test->cmd = TestCleanup;
+    sync_with_supervisor(test_svc);
+}
+
+/*******************************************************************************
+ * TEST SUPERVISOR FUNCTIONS
+ ******************************************************************************/
+
+/** \brief Print the individual test result to the console.
+ */
+static void print_test_result(utest_t *test) {
+    printf("%d.%lu: ", test->pid, test->current);
+    switch (test->result) {
+        case Passed:
+            puts("[✓]");
```

Comment:
My initial intent was to avoid printf as much as possible due to how expensive it is (vs puts which does no string-parsing), but here since I'm doing printf anyway, I could just move it into the switch-case statement. I think the printf statement was added in later.